accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,254 @@
+# Apple Silicon User Experience
+
+Apple has [announced plans to transition its MacOS hardware line]( https://www.apple.com/newsroom/2020/06/apple-announces-mac-transition-to-apple-silicon/) to a new Arm64-based chip that they refer to as “Apple Silicon”.
+
+The transition has a few impacts on the .NET user experience.
+
+## Background
+
+### Apple Multiple Architecture Support
+
+#### Rosetta 2 x64 Emulation
+
+As part of the their transition plan, Apple created the Rosetta 2 `x86_64` emulator to run `x64` code on an Apple Silicon processor. It is intended to support most `x64` code on Apple Silicon without modification.
+
+Usage is generally transparent to the user. When creating a new process, the current process architectural affinity is used to select the best architecture to use for the new process.
+
+The user also has explicit architecture control through the use of the `arch` utility. The user can select a specific architecture when executing a command from the command line i.e. `arch -arch x86_64 <command>`.
+
+While Rosetta 2 is supported for running `x64` code on Apple Silicon, the reverse emulator is not currently shipping.
+
+#### Universal Binaries
+
+For macOS, universal are the default approach. Most binaries are multi-arch. For single arch binaries the kernel can detect and automatically switch the process to run in the correct architecture, either native or Rosetta 2 emulation of x64.
+
+A universal binary can be created simply by combining two or more architecture specific binaries into a universal binary using the `lipo` tool. The process is reversible too. A universal binary can be `thinned` to extract a single arch binaries.
+
+Given both architectures, universal binaries are relatively easy to create. The general syntax would be `lipo <binaryArchA> <binaryArchB> -create -output <binaryUniversal>`
+
+#### Apple Store requires universal binaries
+
+I have heard that publishing native Apple Silicon (or Apps supporting multiple architectures) apps to the Apple store will require them to be published as Universal binaries. I have not been able to find public documentation of this, so the details are not clear.
```

Comment:
@rolfbjarne @marek-safar Is this requirement documented anywhere?

Reply:
I haven't heard anything about it, but it wouldn't surprise me if Apple made it required at some point.

Reply:
AFAIK it's the default setup for xcode projects but still a recommendation. I agree with Rolf that this can change in the future.

Reply:
not yet formally required by Apple, but something we fully expect to be required later this year. As an example, Apple started mandating that all apps have 64-bit slices in their universal binaries about a year before they started telling people that 32-bit was ending. We know they will stop building Intel-based Macs within the next 18 months or so, so we expect them to start requiring Arm slices this summer.